### PR TITLE
Commit to Push In A New Pulse Analysis Processor

### PR DESCRIPTION
**Pull request by rodwyer100**
### Core Files to be Submitted

SvtRawDataAnaProcessor.cxx
SvtRawDataAnaProcessor.h
config/rawSvtAna_cfg.json and config/rawSvtAna_cfgChannel.json
\analysis\selections\svtHit directory and all files contained therein.
\analysis\plotconfigs\rawSvtAna...3.json
\analysis\src\RawSvtHitHistos.cxx

The SvtRawDataAnaProcessor.cxx takes in many cuts on events based on amplitude, time, channel, phase, etc. and handles them using the region_selector object of previous processors. What cuts are meant to be applied are placed in the _cfg (_cfgChannel implements a more aggresive cut that pares down channels; it is separate only so that we don't misapply it) After an event passes these cuts, it checks whether another private int sample_ is enabled. 

Say it is enabled. Then the following occurs:

We give rawSvtAna_cfg a basline and timeProfiles file directory. It expects to find raw dat files from our database including baselines and channel time profiles for some run (14552 in all code currently). Afterwards it populated a private array of the SvtRawDataAnaProcessor object (something like baseLines1[][][][] or baseLines2) with all the basline information for each channel separated into feb and hybrid. baseLines1 handles the first two layers, the other the remaining ones. It also does the same for the characteristic times. Once these are populated, the processor generates root plots of fits for 200 fits among the cuts that a person establishes. For instance, if the processor is given a json file in svtHit with a high chi_sqr threshold cut, then it only plot really nice curves. These curves are just stored wherever you are running the hpstr job

Now that it has performed the sampling pulse profile operation, it moves onto the next step. It generates histograms using the histogram multiplier key as implemented by HistoManager (it makes a separate histogram for each layer and hybrid). The histograms it makes are specified in the rawSvtAna3.json file in analsys/plotconfigs. RawSvtHitHistos.cxx then populates these one and two dimensional histograms. It is here where the time difference between two pulses, etc. is implemented.

##I have verified that this processor (and its adjacent files) work as intended. 

##They will require structure and comment improvements that I intend on implementing either here or as I go on.

### CODE REVIEW

File: analysis/src/HistoManager.cxx [hunk 1]
```diff
@@ -95,6 +95,10 @@ void HistoManager::DefineHistos(){
 void HistoManager::DefineHistos(std::vector<std::string> histoCopyNames, std::string makeCopyJsonTag){
     if (debug_ > 0) std::cout << "[HistoManager] DefineHistos" << std::endl;
     std::string h_name = "";
+    for (auto hist : _h_configs.items()){
```
Comment (cbravo135) [inline analysis/src/HistoManager.cxx]: please remove this loop, or at least require debug_ > 0 for it.

File: analysis/src/HistoManager.cxx [hunk 2]
```diff
@@ -339,7 +345,7 @@ void HistoManager::loadHistoConfig(const std::string histoConfigFile) {
 
     std::ifstream i_file(histoConfigFile);
     i_file >> _h_configs;
-    if (debug_) {
+    if (true) {
```
Comment (cbravo135) [inline analysis/src/HistoManager.cxx]: put debug_ back...

File: analysis/src/HistoManager.cxx [hunk 3]
```diff
@@ -351,23 +357,24 @@ void HistoManager::loadHistoConfig(const std::string histoConfigFile) {
 
 
 void HistoManager::saveHistos(TFile* outF,std::string folder) {
-
+    std::cout<<"hello34"<<std::endl;
```
Comment (cbravo135) [inline analysis/src/HistoManager.cxx]: get rid of all these hellos

File: plotUtils/structFind6.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
+import ROOT as r
```
Comment (cbravo135) [inline plotUtils/structFind6.py]: I think this is in another PR, where I asked you to rename the file to be more descriptive of what the script does.

File: processing/include/ConfigurePython.h [hunk 5]
```diff
@@ -61,6 +61,9 @@ class ConfigurePython {
          */
         int run_mode_{-1};
 
+        /** The number of events to skip. */
```
Comment (cbravo135) [inline processing/include/ConfigurePython.h:64]: why did you need to add this here?

File: processing/include/Process.h [hunk 6]
```diff
@@ -71,13 +71,28 @@ class Process {
         }
 
         /**
+<<<<<<< HEAD
```
Comment (cbravo135) [inline processing/include/Process.h]: you have markings here from a manual overlap merge. When did you add skip events like this?

File: processing/src/HpsEventFile.cxx [hunk 7]
```diff
@@ -38,4 +38,5 @@ bool HpsEventFile::nextEvent() {
   intree_->GetEntry(entry_++);
   
   return true;
-} 
+
```
Comment (cbravo135) [inline processing/src/HpsEventFile.cxx]: why?

File: processors/config/rawSvtAnaHits_cfg.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+rawAnaSvt = HpstrConf.Processor('svtana','SvtRawDataAnaProcessor')
+
+###############################
+#   Processor Configuration   #
+###############################
+
+#SvtRawAnaData
+
+rawAnaSvt.parameters["debug"] = 1
```
Comment (cbravo135) [inline processors/config/rawSvtAnaHits_cfg.py]: we should set debug to 0 by default

File: processors/include/SvtRawDataAnaProcessor.h [hunk 9]
```diff
@@ -0,0 +1,101 @@
+#ifndef __RAWSVTHIT_ANAPROCESSOR_H__
+#define __RAWSVTHIT_ANAPROCESSOR_H__
+
+//HPSTR
+#include "HpsEvent.h"
+#include "RawSvtHit.h"
+#include "RawSvtHitHistos.h"
+#include "AnaHelpers.h"
+#include "Event.h"
+#include "BaseSelector.h"
```
Comment (cbravo135) [inline processors/include/SvtRawDataAnaProcessor.h:10]: So you don't use BaseSelectorSVT? Why did you add that class? Probably remove it, looks the same as BaseSelector to me anyway. You can always just inherit from BaseSelector.

File: processors/src/SvtRawDataProcessor.cxx [hunk 10]
```diff
@@ -138,7 +138,7 @@ bool SvtRawDataProcessor::process(IEvent* ievent) {
 
             rawHit->setFitN(rawTracker_hit_fits_list.size());
         }
-        rawhits_.push_back(rawHit);
+       rawhits_.push_back(rawHit);
```
Comment (cbravo135) [inline processors/src/SvtRawDataProcessor.cxx]: The white space was more correct without this change.

File: processors/include/SvtRawDataAnaProcessor.h (hunk 9 above)
Comment (rodwyer100) [reply processors/include/SvtRawDataAnaProcessor.h:10]: Removed.

File: processors/config/rawSvtAnaHits_cfg.py (hunk 8 above)
Comment (rodwyer100) [reply processors/config/rawSvtAnaHits_cfg.py]: Done

File: analysis/src/HistoManager.cxx (hunk 1 above)
Comment (rodwyer100) [reply analysis/src/HistoManager.cxx]: Done

File: processors/src/SvtRawDataProcessor.cxx (hunk 10 above)
Comment (rodwyer100) [reply processors/src/SvtRawDataProcessor.cxx]: Fixed Again

File: processing/src/HpsEventFile.cxx (hunk 7 above)
Comment (rodwyer100) [reply processing/src/HpsEventFile.cxx]: Accident. Was fixed

File: processing/include/Process.h (hunk 6 above)
Comment (rodwyer100) [reply processing/include/Process.h]: I added it near the beginning. I wanted to use skip events to perform debugging.

File: processing/include/ConfigurePython.h (hunk 5 above)
Comment (rodwyer100) [reply processing/include/ConfigurePython.h:64]: To perform debugging

File: plotUtils/structFind6.py (hunk 4 above)
Comment (rodwyer100) [reply plotUtils/structFind6.py]: I will remove this from this current PR

File: analysis/src/HistoManager.cxx (hunk 2 above)
Comment (rodwyer100) [reply analysis/src/HistoManager.cxx]: Fixed.

File: analysis/src/HistoManager.cxx (hunk 3 above)
Comment (rodwyer100) [reply analysis/src/HistoManager.cxx]: Done dizzle pho shizzle.

**Comment by rodwyer100** [comment]
I have fiddled again. Let me know what you think.

File: analysis/include/RawSvtHitHistos2.h [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+//#include "BaseSelector.h"
+//#include "AnaHelpers.h"
+
+#include "ModuleMapper.h"
+
+#include <string>
+
+
+class RawSvtHitHistos2 : public HistoManager{
```
Comment (cbravo135) [inline analysis/include/RawSvtHitHistos2.h]: Why are there two RawSvtHitHistos classes?

Comment (rodwyer100) [reply analysis/include/RawSvtHitHistos2.h]: Was a very early thing that I was doing to get RawSvtHitAnaProcessor working. Never needed it. Has been removed.

Comment (rodwyer100) [reply analysis/include/RawSvtHitHistos2.h]: I believe that that is the only such instance but will read it over carefully.

File: analysis/plotconfigs/svt/rawSvtAnaHits2.json [hunk 12]
```diff
@@ -0,0 +1,812 @@
+{
```
Comment (cbravo135) [inline analysis/plotconfigs/svt/rawSvtAnaHits2.json]: Why are there 3 different version of this with just a difference of a number in the filename?

File: analysis/src/HistoManager.cxx [hunk 13]
```diff
@@ -108,7 +114,8 @@ void HistoManager::DefineHistos(std::vector<std::string> histoCopyNames, std::st
             std::cout << "DefineHisto: " << h_name << std::endl;
             std::size_t found = (hist.key()).find_last_of("_");
             std::string extension = hist.key().substr(found+1);
-
+            std::string xtitty = hist.value().at("xtitle");
```
Comment (cbravo135) [inline analysis/src/HistoManager.cxx]: please put this printout of the histogram xtitles inside of an if(debug_>0)

File: processing/src/ConfigurePython.cxx [hunk 14]
```diff
@@ -129,6 +129,7 @@ ConfigurePython::ConfigurePython(const std::string& python_script, char* args[],
 
     event_limit_ = intMember(p_process, "max_events");
     run_mode_    = intMember(p_process, "run_mode");
+    skip_events_    = intMember(p_process, "skip_events");
```
Comment (cbravo135) [inline processing/src/ConfigurePython.cxx:132]: Adding this here will mean every config file will require skip_events to be set. Please go through all of them and double check they all have it set.

File: processors/config/rawSvtAnaHits_cfg.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+root2_file = options.outFilename
+
+print('Root file Input: %s' % root1_file)
+print('Root file Output: %s' % root2_file)
+
+p = HpstrConf.Process()
+
+p.run_mode = 1
+p.skip_events = 0
+p.max_events = 200000
```
Comment (cbravo135) [inline processors/config/rawSvtAnaHits_cfg.py]: It would be better to add these as options to the arg parser instead of hard coding them like this.

File: processing/src/HpsEventFile.cxx [hunk 16]
```diff
@@ -38,4 +38,4 @@ bool HpsEventFile::nextEvent() {
   intree_->GetEntry(entry_++);
   
   return true;
-} 
+}
```
Comment (cbravo135) [inline processing/src/HpsEventFile.cxx]: There is no need to change the whitespace of HpsEventFile.h or HpsEventFile.cxx in this PR. Can you change them back? If you want to adjust the whitespace, please do this consistently across the whole project in a single PR. Please collaborate with Sarah in doing this, as she has already been thinking about linting in general.

File: analysis/src/HistoManager.cxx (hunk 13 above)
Comment (rodwyer100) [reply analysis/src/HistoManager.cxx]: Done

File: analysis/plotconfigs/svt/rawSvtAnaHits2.json (hunk 12 above)
Comment (rodwyer100) [reply analysis/plotconfigs/svt/rawSvtAnaHits2.json]: I have left the non obsolete one. Renamed it without the numbering. The numbers referred to different versions with different diagnostic plots. It seems like they were just different bounds on the plots; I haven't used earlier ones in eons.

File: processing/src/HpsEventFile.cxx (hunk 16 above)
Comment (rodwyer100) [reply processing/src/HpsEventFile.cxx]: I have not intentionally changed the whitespace. I have changed things back. I think Sarah and I will handle that in an iteration on this push request; may be a bit before we have a long time just to talk.

File: processors/config/rawSvtAnaHits_cfg.py (hunk 15 above)
Comment (rodwyer100) [reply processors/config/rawSvtAnaHits_cfg.py]: Did it

File: processing/src/ConfigurePython.cxx (hunk 14 above)
Comment (rodwyer100) [reply processing/src/ConfigurePython.cxx:132]: Did this

**Comment by rodwyer100** [comment]
Alright made all the changes required. Let me know if I put too many skip statements; I think it should be correct.

**Comment by rodwyer100** [comment]
I have not fixed the simple branch conflicts but I think they can be just changed to my version. Let me know if you disagree.

File: processing/include/HpsEventFile.h [hunk 17]
```diff
@@ -60,7 +60,7 @@ class HpsEventFile : public IEventFile {
         void close();
 
 
-    private:
+    private:
```
Comment (cbravo135) [inline processing/include/HpsEventFile.h]: I don't think this whitespace change is needed.

File: processors/config/anaKalSimpTuple_cfg.py [hunk 18]
```diff
@@ -105,8 +105,9 @@
     print ("ERROR::Need to specify which tracks KF or GBL")
     exit(1)
 
-if (options.nevents > 0):
-    p.max_events = options.nevents
+p.skip_events = options.skip_events
+if(options.nevents>-1):
```
Comment (cbravo135) [inline processors/config/anaKalSimpTuple_cfg.py]: You are not setting p.max_events when options.nevents = -1

File: processors/config/rawSvtAnaHits_cfg.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+
+rawAnaSvt.parameters["debug"] = 0
+rawAnaSvt.parameters["anaName"] = 'rawSvtHitAna'
+rawAnaSvt.parameters["trkrHitColl"] = 'SVTRawHitsOnTrack_KF'#'SVTRawTrackerHits'
+rawAnaSvt.parameters["histCfg"] = os.environ['HPSTR_BASE']+'/analysis/plotconfigs/svt/rawSvtAnaHits.json'
+rawAnaSvt.parameters["sample"] = 0 
+
+RegionPath = os.environ['HPSTR_BASE']+"/analysis/selections/svtHit/"
+
+rawAnaSvt.parameters["baselineFile"]="/sdf/group/hps/users/rodwyer1/hps_14552_offline_baselines.dat"
```
Comment (cbravo135) [inline processors/config/rawSvtAnaHits_cfg.py]: These shouldn't be hard coded paths to your user dir on sdf. We should add these dat files to hpstr and you should use the path relative to HPSTR_BASE.

File: analysis/src/HistoManager.cxx [hunk 20]
```diff
@@ -104,11 +110,15 @@ void HistoManager::DefineHistos(std::vector<std::string> histoCopyNames, std::st
...
-            std::cout << "DefineHisto: " << h_name << std::endl;
-            std::size_t found = (hist.key()).find_last_of("_");
-            std::string extension = hist.key().substr(found+1);
-
+            if(debug_){
+                std::cout << "DefineHisto: " << h_name << std::endl;
+            }
+                std::size_t found = (hist.key()).find_last_of("_");
+                std::string extension = hist.key().substr(found+1);
+                std::string xtitty = hist.value().at("xtitle");
```
Comment (cbravo135) [inline analysis/src/HistoManager.cxx]: xtitty??

Comment (cbravo135) [reply analysis/src/HistoManager.cxx]: since "extension" and "xtitty" are only used here in the debug, maybe we should not make variables for them and just call these in the std::cout stream line.

File: analysis/plotconfigs/svt/Svt2DBl.json [hunk 21]
```diff
@@ -8,20 +8,11 @@
...
-            "ytitle" : "Events"
-
-        },
-
     "SvtHybridsHitN_h" : {
             "xtitle" : "# of Strips",
             "bins" : 10000,
             "minX" : 0,
-            "maxX" : 10000,
+            "maxX" : 200,
```
Comment (cbravo135) [inline analysis/plotconfigs/svt/Svt2DBl.json]: this should be higher than 200, but also there is no reason to have more bins than integers on the x axis. make this something like 2000 bins from -0.5 to 1999.5

File: processors/config/anaKalSimpTuple_cfg.py (hunk 18 above)
Comment (rodwyer100) [reply processors/config/anaKalSimpTuple_cfg.py]: I think if I dont set it, it will run over everything automatically

File: processors/config/rawSvtAnaHits_cfg.py (hunk 19 above)
Comment (rodwyer100) [reply processors/config/rawSvtAnaHits_cfg.py]: Done :)

File: analysis/plotconfigs/svt/Svt2DBl.json (hunk 21 above)
Comment (rodwyer100) [reply analysis/plotconfigs/svt/Svt2DBl.json]: Done

File: processing/include/HpsEventFile.h (hunk 17 above)
Comment (rodwyer100) [reply processing/include/HpsEventFile.h]: I promise these aren't intentional.

File: analysis/src/HistoManager.cxx (hunk 20 above)
Comment (rodwyer100) [reply analysis/src/HistoManager.cxx]: extension is not just used for debug; it also sets up whether we populate a 1 or 2d histogram. title may be only used for debug. So it has been removed

**Comment by rodwyer100** [comment]
Found some files that were untracked. Should have all your changes implemented now, and included the dat files locally

File: analysis/src/HistoManager.cxx [hunk 22]
```diff
@@ -104,11 +110,14 @@ void HistoManager::DefineHistos(std::vector<std::string> histoCopyNames, std::st
                 h_name = m_name+"_"+ histoCopyNames.at(i) + "_" + hist.key() ;
                 singleCopy = false;
             }
-
-            std::cout << "DefineHisto: " << h_name << std::endl;
-            std::size_t found = (hist.key()).find_last_of("_");
-            std::string extension = hist.key().substr(found+1);
-
+            if(debug_){
```
Comment (cbravo135) [inline analysis/src/HistoManager.cxx:115]: We might as well just have one if(debug_) in this section of the code, and just cout all the stuff in that one, rather than the two.

File: analysis/include/BaseSelectorSVT.h [hunk 23]
```diff
@@ -0,0 +1,59 @@
+#ifndef BASESELECTOR_H
```
Comment (cbravo135) [inline analysis/include/BaseSelectorSVT.h]: Why is this whole new class needed? This looks like a copy of BaseSelector.h, but without the nice comments for the Doxygen documentation. Is it possible to achieve this class by inheriting from BaseSelector? It is unclear to me what this has that BaseSelector does not.

File: processing/src/HpsEventFile.cxx [hunk 24]
```diff
@@ -38,4 +38,4 @@ bool HpsEventFile::nextEvent() {
...
```
Comment (cbravo135) [inline processing/src/HpsEventFile.cxx:41]: I don't think we need to change the whitespace here in this PR.

File: analysis/src/HistoManager.cxx (hunk 22 above)
Comment (rodwyer100) [reply analysis/src/HistoManager.cxx:115]: Done.

File: analysis/include/BaseSelectorSVT.h (hunk 23 above)
Comment (rodwyer100) [reply analysis/include/BaseSelectorSVT.h]: Fixed. It again was an early artefact of getting things working. I had deleted the cpp file, and forgotten to remove the corresponding h file. Have done now.

File: processing/src/HpsEventFile.cxx (hunk 24 above)
Comment (rodwyer100) [reply processing/src/HpsEventFile.cxx:41]: I don't honestly don't know why it keeps saying I fiddled with the whitespace. This is not intentional. I changed it back I think from the previous whitespace. I promise we'll adopt a convention when I talk with Sarah.

**Comment by omar-moreno** [comment]
Before you commit just run:

find . -type f -name '*.h' -o -name '*.cxx' -exec clang-format -i
--style=Google {} ';'

This will give you a consistent style.

**Comment by rodwyer100** [comment]
I don't think I have the clang-format command installed.

**Comment by rodwyer100** [comment]
*and I don't have sudo privileges. I could try this locally.

**Comment by rodwyer100** [comment]
I made a commit with the style changes. For some reason it didn't let me make a git message.

**Comment by cbravo135** [comment]
I had requested in a previous comment we wait until we have a discussion on what style to use to change style, and do it in a separate PR. Can you revert that commit that changed the style and open a separate PR with this?

**Comment by rodwyer100** [comment]
Done

File: processors/config/cnvStdhepToRoot_cfg.py [hunk 25]
```diff
@@ -29,7 +31,7 @@
 #MCParticles
 cnvStd.parameters["mcPartCollStdhep"] = 'MCParticle'
 cnvStd.parameters["mcPartCollRoot"] = 'MCParticle'
-cnvStd.parameters["maxEvent"] = options.nevents
+cnvStd.parameters["maxEvent"] = options.skip_events+options.nevents
```
Comment (cbravo135) [inline processors/config/cnvStdhepToRoot_cfg.py]: This won't work if options.nevents is -1, which is pretty typical.

Comment (rodwyer100) [reply processors/config/cnvStdhepToRoot_cfg.py]: I had to do change three files to get a consistent thing. Let me know if you think I've done it right.

File: analysis/plotconfigs/svt/Svt2DBl.json [hunk 26]
```diff
@@ -8,20 +8,11 @@
...
-
-        },
-
     "SvtHybridsHitN_h" : {
             "xtitle" : "# of Strips",
-            "bins" : 10000,
-            "minX" : 0,
-            "maxX" : 10000,
+            "bins" : 2000,
+            "minX" : -.5,
```
Comment (cbravo135) [inline analysis/plotconfigs/svt/Svt2DBl.json]: -.5 is gonna crash. Change this to -0.5

File: processors/include/SvtRawDataAnaProcessor.h [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+        std::string svtHitColl_{"RotatedHelicalTrackHits"};
+        std::vector<std::string> regionSelections_;
+        std::map<std::string, std::shared_ptr<BaseSelector>> reg_selectors_;
+        std::map<std::string,std::shared_ptr<RawSvtHitHistos>> reg_histos_;
+        typedef std::map<std::string,std::shared_ptr<RawSvtHitHistos>>::iterator reg_it;
+        std::vector<std::string> regions_;
+        std::string baselineFile_;
+        std::string timeProfiles_;
+        
+        //I WILL FIX THIS QUICKLY JUST INPUT WHAT YOU HAVE IN THE PYTHON FILE
```
Comment (cbravo135) [inline processors/include/SvtRawDataAnaProcessor.h]: Is this comment still relevant??

Comment (rodwyer100) [reply processors/include/SvtRawDataAnaProcessor.h]: No I fixed some hard coded thing a while ago. Removed the comment.

**Review by cbravo135 — APPROVED** [review]
Ran hps-mc hpstr example as a test. This adds a great new feature allowing the user to skip a number of events specified at runtime.
